File: javascript/jsfind/src/findoptions.js
/*
 * findoptions.js
 *
 * defines the set of find options and provides functionality to define find settings from them
 */

const config = require('./config');
const {ArgTokenizer} = require('./argtokenizer');
const {ArgTokenType} = require('./argtokentype')
const {FileUtil} = require('./fileutil');
const {FindError} = require('./finderror');
const {FindOption} = require('./findoption');
const {FindSettings} = require('./findsettings');
const {nameToSortBy} = require("./sortby");

class FindOptions {
    constructor() {
        this.boolActionMap = {
            'archivesonly':
              (b, settings) => { settings.archivesOnly = b; },
            'colorize':
              (b, settings) => { settings.colorize = b; },
            'debug':
              (b, settings) => { settings.debug = b; },
            'excludearchives':
              (b, settings) => { settings.includeArchives = !b; },
            'excludehidden':
              (b, settings) => { settings.includeHidden = !b; },
            'followsymlinks':
              (b, settings) => { settings.followSymlinks = b; },
            'help':
              (b, settings) => { settings.printUsage = b; },
            'includearchives':
              (b, settings) => { settings.includeArchives = b; },
            'includehidden':
              (b, settings) => { settings.includeHidden = b; },
            'nocolorize':
              (b, settings) => { settings.colorize = !b; },
            'nofollowsymlinks':
              (b, settings) => { settings.followSymlinks = !b; },
            'noprintdirs':
              (b, settings) => { settings.printDirs = !b; },
            'noprintfiles':
              (b, settings) => { settings.printFiles = !b; },
            'norecursive':
              (b, settings) => { settings.recursive = !b; },
            'printdirs':
              (b, settings) => { settings.printDirs = b; },
            'printfiles':
              (b, settings) => { settings.printFiles = b; },
            'recursive':
              (b, settings) => { settings.recursive = b; },
            'sort-ascending':
              (b, settings) => { settings.sortDescending = !b; },
            'sort-caseinsensitive':
              (b, settings) => { settings.sortCaseInsensitive = b; },
            'sort-casesensitive':
              (b, settings) => { settings.sortCaseInsensitive = !b; },
            'sort-descending':
              (b, settings) => { settings.sortDescending = b; },
            'verbose':
              (b, settings) => { settings.verbose = b; },
            'version':
              (b, settings) => { settings.printVersion = b; }
        };
        this.stringActionMap = {
            'in-archiveext':
              (s, settings) => { settings.addInArchiveExtensions(s); },
            'in-archivefilepattern':
              (s, settings) => { settings.addInArchiveFilePatterns(s); },
            'in-dirpattern':
              (s, settings) => { settings.addInDirPatterns(s); },
            'in-ext':
              (s, settings) => { settings.addInExtensions(s); },
            'in-filepattern':
              (s, settings) => { settings.addInFilePatterns(s); },
            'in-filetype':
              (s, settings) => { settings.addInFileTypes(s); },
            'maxlastmod':
              (s, settings) => { settings.maxLastModFromString(s); },
            'minlastmod':
              (s, settings) => { settings.minLastModFromString(s); },
            'out-dirpattern':
              (s, settings) => { settings.addOutDirPatterns(s); },
            'out-archiveext':
              (s, settings) => { settings.addOutArchiveExtensions(s); },
            'out-archivefilepattern':
              (s, settings) => { settings.addOutArchiveFilePatterns(s); },
            'out-ext':
              (s, settings) => { settings.addOutExtensions(s); },
            'out-filepattern':
              (s, settings) => { settings.addOutFilePatterns(s); },
            'out-filetype':
              (s, settings) => { settings.addOutFileTypes(s); },
            'path':
              (s, settings) => { settings.paths.push(s); },
            'sort-by':
              (s, settings) => { settings.sortBy = nameToSortBy(s); }
        };
        this.intActionMap = {
            'maxdepth':
              (i, settings) => { settings.maxDepth = i; },
            'maxsize':
              (i, settings) => { settings.maxSize = i; },
            'mindepth':
              (i, settings) => { settings.minDepth = i; },
            'minsize':
              (i, settings) => { settings.minSize = i; },
        };

        // the list of FindOption objects (populated from JSON)
        this.options = [];

        // populate options from JSON file
        (() => {
            let json = FileUtil.getFileContentsSync(config.FIND_OPTIONS_JSON_PATH, 'utf-8');
            let obj = JSON.parse(json);
            if (Object.prototype.hasOwnProperty.call(obj, 'findoptions') && Array.isArray(obj.findoptions)) {
                obj.findoptions.forEach(fo => {
                    let longArg = fo.long;
                    let shortArg = '';
                    if (Object.prototype.hasOwnProperty.call(fo, 'short')) {
                        shortArg = fo.short;
                    }
                    let desc = fo.desc;
                    let argType = ArgTokenType.Unknown;
                    if (this.boolActionMap[longArg]) {
                        argType = ArgTokenType.Bool;
                    } else if (this.stringActionMap[longArg]) {
                        argType = ArgTokenType.Str;
                    } else if (this.intActionMap[longArg]) {
                        argType = ArgTokenType.Int;
                    }
                    this.options.push(new FindOption(shortArg, longArg, desc, argType));
                });
            } else throw new FindError(`Invalid findoptions file: ${config.FIND_OPTIONS_JSON_PATH}`);
            this.argTokenizer = new ArgTokenizer(this.options);
        })();
    }

    updateSettingsFromArgTokens(settings, argTokens) {
        let err = null;
        for (const argToken of argTokens) {
            if (err) break;
            if (argToken.type === ArgTokenType.Bool) {
                if (typeof argToken.value === 'boolean') {
                    this.boolActionMap[argToken.name](argToken.value, settings);
                } else {
                    err = new FindError(`Invalid value for option: ${argToken}`);
                }
            } else if (argToken.type === ArgTokenType.Str) {
                if (argToken.name === 'settings-file') {
                    err = this.updateSettingsFromFile(settings, argToken.value);
                } else if (typeof argToken.value === 'string') {
                    this.stringActionMap[argToken.name](argToken.value, settings);
                } else if (typeof argToken.value === 'object' && Array.isArray(argToken.value)) {
                    argToken.value.forEach(s => {
                        if (typeof s === 'string') {
                            this.stringActionMap[argToken.name](s, settings);
                        } else {
                            err = new FindError(`Invalid value for option: ${argToken}`);
                        }
                    });
                } else {
                    err = new FindError(`Invalid value for option: ${argToken}`);
                }
            } else if (argToken.type === ArgTokenType.Int) {
                if (typeof argToken.value === 'number') {
                    this.intActionMap[argToken.name](argToken.value, settings);
                } else {
                    err = new FindError(`Invalid value for option: ${argToken}`);
                }
            } else {
                err = new FindError(`Invalid option: ${argToken}`);
            }
        }
        return err;
    }

    updateSettingsFromJson(settings, json) {
        let { err, argTokens } = this.argTokenizer.tokenizeJson(json);
        if (!err) {
            err = this.updateSettingsFromArgTokens(settings, argTokens);
        }
        return err;
    }

    updateSettingsFromFile(settings, filePath) {
        let { err, argTokens } = this.argTokenizer.tokenizeFile(filePath);
        if (!err) {
            err = this.updateSettingsFromArgTokens(settings, argTokens);
        }
        return err;
    }

    updateSettingsFromArgs(settings, args) {
        let { err, argTokens } = this.argTokenizer.tokenizeArgs(args);
        if (!err) {
            err = this.updateSettingsFromArgTokens(settings, argTokens);
        }
        return err;
    }

    settingsFromArgs(args, cb) {
        let settings = new FindSettings();
        // default printFiles to true since running as cli
        settings.printFiles = true;
        let err = this.updateSettingsFromArgs(settings, args);
        cb(err, settings);
    }

    optCmp(o1, o2) {
        const a = o1.sortArg;
        const b = o2.sortArg;
        return a.localeCompare(b);
    }

    getUsageString() {
        let usage = 'Usage:\n jsfind [options] <path> [<path> ...]\n\n';
        usage += 'Options:\n';
        let optStrings = [];
        let optDescs = [];
        let longest = 0;
        this.options.sort(this.optCmp);
        this.options.forEach(opt => {
            let optString = ' ';
            if (opt.shortArg)
                optString += '-' + opt.shortArg + ',';
            optString += '--' + opt.longArg;
            if (optString.length > longest)
                longest = optString.length;
            optStrings.push(optString);
            optDescs.push(opt.desc);
        });
        for (let i=0; i < optStrings.length; i++) {
            let os = optStrings[i];
            while (os.length < longest)
                os += ' ';
            usage += os + '  ' + optDescs[i] + '\n';
        }
        return usage;
    }

    usageWithCode(exitCode) {
        console.log(this.getUsageString());
        process.exit(exitCode);
    }

    usage() {
        this.usageWithCode(0);
    }
}

exports.FindOptions = FindOptions;

File: go/gofind/pkg/gofind/findoptions.go
package gofind

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

type FindOption struct {
	Short   string
	Long    string
	Desc    string
	argType ArgTokenType
}

func (o FindOption) ShortArg() string {
	return o.Short
}

func (o FindOption) LongArg() string {
	return o.Long
}

func (o FindOption) Description() string {
	return o.Desc
}

func (o FindOption) ArgType() ArgTokenType {
	return o.argType
}

type FindOptions struct {
	FindOptions     []*FindOption
	BoolActionMap   map[string]boolAction
	StringActionMap map[string]stringAction
	IntActionMap    map[string]intAction
	LongActionMap   map[string]longAction
	ArgTokenizer    *ArgTokenizer
}

type boolAction func(b bool, settings *FindSettings)

func getBoolActionMap() map[string]boolAction {
	m := map[string]boolAction{
		"archivesonly": func(b bool, settings *FindSettings) {
			settings.SetArchivesOnly(b)
		},
		"colorize": func(b bool, settings *FindSettings) {
			settings.SetColorize(b)
		},
		"debug": func(b bool, settings *FindSettings) {
			settings.SetDebug(b)
		},
		"excludearchives": func(b bool, settings *FindSettings) {
			settings.SetIncludeArchives(!b)
		},
		"excludehidden": func(b bool, settings *FindSettings) {
			settings.SetIncludeHidden(!b)
		},
		"followsymlinks": func(b bool, settings *FindSettings) {
			settings.SetFollowSymlinks(b)
		},
		"help": func(b bool, settings *FindSettings) {
			settings.SetPrintUsage(b)
		},
		"includearchives": func(b bool, settings *FindSettings) {
			settings.SetIncludeArchives(b)
		},
		"includehidden": func(b bool, settings *FindSettings) {
			settings.SetIncludeHidden(b)
		},
		"nocolorize": func(b bool, settings *FindSettings) {
			settings.SetColorize(!b)
		},
		"nofollowsymlinks": func(b bool, settings *FindSettings) {
			settings.SetFollowSymlinks(!b)
		},
		"noprintdirs": func(b bool, settings *FindSettings) {
			settings.SetPrintDirs(!b)
		},
		"noprintfiles": func(b bool, settings *FindSettings) {
			settings.SetPrintFiles(!b)
		},
		"norecursive": func(b bool, settings *FindSettings) {
			settings.SetRecursive(!b)
		},
		"printdirs": func(b bool, settings *FindSettings) {
			settings.SetPrintDirs(b)
		},
		"printfiles": func(b bool, settings *FindSettings) {
			settings.SetPrintFiles(b)
		},
		"recursive": func(b bool, settings *FindSettings) {
			settings.SetRecursive(b)
		},
		"sort-ascending": func(b bool, settings *FindSettings) {
			settings.SetSortDescending(!b)
		},
		"sort-caseinsensitive": func(b bool, settings *FindSettings) {
			settings.SetSortCaseInsensitive(b)
		},
		"sort-casesensitive": func(b bool, settings *FindSettings) {
			settings.SetSortCaseInsensitive(!b)
		},
		"sort-descending": func(b bool, settings *FindSettings) {
			settings.SetSortDescending(b)
		},
		"verbose": func(b bool, settings *FindSettings) {
			settings.SetVerbose(b)
		},
		"version": func(b bool, settings *FindSettings) {
			settings.SetPrintVersion(b)
		},
	}
	return m
}

type stringAction func(s string, settings *FindSettings)

func getStringActionMap() map[string]stringAction {
	m := map[string]stringAction{
		"in-archiveext": func(s string, settings *FindSettings) {
			settings.AddInArchiveExtension(s)
		},
		"in-archivefilepattern": func(s string, settings *FindSettings) {
			settings.AddInArchiveFilePattern(s)
		},
		"in-dirpattern": func(s string, settings *FindSettings) {
			settings.AddInDirPattern(s)
		},
		"in-ext": func(s string, settings *FindSettings) {
			settings.AddInExtension(s)
		},
		"in-filepattern": func(s string, settings *FindSettings) {
			settings.AddInFilePattern(s)
		},
		"in-filetype": func(s string, settings *FindSettings) {
			settings.AddInFileType(GetFileTypeForName(s))
		},
		"maxlastmod": func(s string, settings *FindSettings) {
			settings.SetMaxLastModFromString(s)
		},
		"minlastmod": func(s string, settings *FindSettings) {
			settings.SetMinLastModFromString(s)
		},
		"out-archiveext": func(s string, settings *FindSettings) {
			settings.AddOutArchiveExtension(s)
		},
		"out-archivefilepattern": func(s string, settings *FindSettings) {
			settings.AddOutArchiveFilePattern(s)
		},
		"out-dirpattern": func(s string, settings *FindSettings) {
			settings.AddOutDirPattern(s)
		},
		"out-ext": func(s string, settings *FindSettings) {
			settings.AddOutExtension(s)
		},
		"out-filepattern": func(s string, settings *FindSettings) {
			settings.AddOutFilePattern(s)
		},
		"out-filetype": func(s string, settings *FindSettings) {
			settings.AddOutFileType(GetFileTypeForName(s))
		},
		"path": func(s string, settings *FindSettings) {
			settings.AddPath(s)
		},
		"sort-by": func(s string, settings *FindSettings) {
			settings.SetSortByFromString(s)
		},
	}
	return m
}

type intAction func(i int, settings *FindSettings)

func getIntActionMap() map[string]intAction {
	m := map[string]intAction{
		"maxdepth": func(i int, settings *FindSettings) {
			settings.SetMaxDepth(i)
		},
		"mindepth": func(i int, settings *FindSettings) {
			settings.SetMinDepth(i)
		},
	}
	return m
}

type longAction func(l int64, settings *FindSettings)

func getLongActionMap() map[string]longAction {
	m := map[string]longAction{
		"maxsize": func(l int64, settings *FindSettings) {
			settings.SetMaxSize(l)
		},
		"minsize": func(l int64, settings *FindSettings) {
			settings.SetMinSize(l)
		},
	}
	return m
}

type JsonFindOptions struct {
	FindOptions []*FindOption
}

func FindOptionsFromJson() (*FindOptions, error) {
	config := NewFindConfig()
	data, err := os.ReadFile(config.FINDOPTIONSPATH)
	if err != nil {
		return &FindOptions{}, err
	}
	var jsonFindOptions JsonFindOptions
	if err = json.Unmarshal(data, &jsonFindOptions); err != nil {
		return &FindOptions{}, err
	}

	// TEMPORARY
	//findOptions.generateCodeFile("/Users/cary/src/xfind/go/gofind/pkg/gofind/findoptionsgen.go")

	var findOptions []*FindOption
	boolActionMap := getBoolActionMap()
	stringActionMap := getStringActionMap()
	intActionMap := getIntActionMap()
	longActionMap := getLongActionMap()

	for _, jo := range jsonFindOptions.FindOptions {
		var argType ArgTokenType
		if _, isBool := boolActionMap[jo.Long]; isBool {
			argType = ArgTokenTypeBool
		} else if _, isString := stringActionMap[jo.Long]; isString {
			argType = ArgTokenTypeString
		} else if _, isInt := intActionMap[jo.Long]; isInt {
			argType = ArgTokenTypeInt
		} else if _, isLong := longActionMap[jo.Long]; isLong {
			argType = ArgTokenTypeLong
		}
		fo := &FindOption{Short: jo.Short, Long: jo.Long, Desc: jo.Desc, argType: argType}
		findOptions = append(findOptions, fo)
	}

	argOptions := make([]ArgOption, len(findOptions))
	for i, o := range findOptions {
		argOptions[i] = o
	}

	return &FindOptions{
		findOptions,
		boolActionMap,
		stringActionMap,
		intActionMap,
		longActionMap,
		NewArgTokenizer(argOptions),
	}, nil
}

func NewFindOptions() *FindOptions {
	findOptions, err := FindOptionsFromJson()
	if err != nil {
		// do something
	}
	return findOptions
}

func (fo *FindOptions) updateSettingsFromArgTokens(settings *FindSettings, argTokens []*ArgToken) error {
	for _, argToken := range argTokens {
		if argToken.Type == ArgTokenTypeBool {
			if bf, isBool := fo.BoolActionMap[argToken.Name]; isBool {
				bf(argToken.Value.(bool), settings)
			} else {
				return fmt.Errorf("Invalid value for option: %v", argToken.Name)
			}
		} else if argToken.Type == ArgTokenTypeString {
			if sf, isString := fo.StringActionMap[argToken.Name]; isString {
				sf(argToken.Value.(string), settings)
			} else if argToken.Name == "settings-file" {
				err := fo.UpdateSettingsFromFile(settings, argToken.Value.(string))
				if err != nil {
					return err
				}
			} else {
				return fmt.Errorf("Invalid value for option: %v", argToken.Name)
			}
		} else if argToken.Type == ArgTokenTypeInt {
			if iff, isInt := fo.IntActionMap[argToken.Name]; isInt {
				iff(argToken.Value.(int), settings)
			} else {
				return fmt.Errorf("Invalid value for option: %v", argToken.Name)
			}
		} else if argToken.Type == ArgTokenTypeLong {
			if lff, isLong := fo.LongActionMap[argToken.Name]; isLong {
				lff(argToken.Value.(int64), settings)
			} else {
				return fmt.Errorf("Invalid value for option: %v", argToken.Name)
			}
		} else {
			return fmt.Errorf("Invalid option: %v", argToken.Name)
		}
	}
	return nil
}

func (fo *FindOptions) UpdateSettingsFromJson(settings *FindSettings, jsonString string) error {
	argTokens, err := fo.ArgTokenizer.TokenizeJson(jsonString)
	if err != nil {
		return err
	}
	return fo.updateSettingsFromArgTokens(settings, argTokens)
}

func (fo *FindOptions) UpdateSettingsFromFile(settings *FindSettings, filePath string) error {
	argTokens, err := fo.ArgTokenizer.TokenizeFile(filePath)
	if err != nil {
		return err
	}
	return fo.updateSettingsFromArgTokens(settings, argTokens)
}

func (fo *FindOptions) UpdateSettingsFromArgs(settings *FindSettings, args []string) error {
	argTokens, err := fo.ArgTokenizer.TokenizeArgs(args)
	if err != nil {
		return err
	}
	return fo.updateSettingsFromArgTokens(settings, argTokens)
}

func (fo *FindOptions) FindSettingsFromArgs(args []string) (*FindSettings, error) {
	settings := GetDefaultFindSettings()
	// default printFiles to true since running as cli
	settings.SetPrintFiles(true)

	err := fo.UpdateSettingsFromArgs(settings, args)

	return settings, err
}

func (fo *FindOptions) getUsageString() string {
	var buffer bytes.Buffer
	buffer.WriteString("\nUsage:\n")
	buffer.WriteString(" gofind [options] <path> [<path> ...]\n\nOptions:\n")
	sortKeyMap := fo.getSortKeyMap()
	optStringMap := fo.getOptStringMap()
	optDescMap := fo.getOptDescMap()
	sortedKeys := GetSortedKeys(sortKeyMap)
	optStrings := GetMapValues(optStringMap)
	longestLen := GetLongestLen(optStrings)
	optFormat := fmt.Sprintf(" %%-%ds  %%s\n", longestLen)
	for _, k := range sortedKeys {
		o := optStringMap[sortKeyMap[k]]
		d := optDescMap[sortKeyMap[k]]
		buffer.WriteString(fmt.Sprintf(optFormat, o, d))
	}
	return buffer.String()
}

func (fo *FindOptions) PrintUsage() {
	Log(fo.getUsageString())
	os.Exit(0)
}

func (fo *FindOptions) PrintVersion() {
	config := NewFindConfig()
	Log(fmt.Sprintf("xfind version %s", config.VERSION))
	os.Exit(0)
}

func (fo *FindOptions) getSortKeyMap() map[string]string {
	m := map[string]string{}
	for _, o := range fo.FindOptions {
		sortKey := ""
		if o.ShortArg() == "" {
			sortKey = strings.ToLower(o.LongArg())
		} else {
			sortKey = fmt.Sprintf("%s@%s", strings.ToLower(o.ShortArg()),
				strings.ToLower(o.LongArg()))
		}
		m[sortKey] = o.LongArg()
	}
	return m
}

func (fo *FindOptions) getOptStringMap() map[string]string {
	m := map[string]string{}
	for _, o := range fo.FindOptions {
		optString := ""
		if o.ShortArg() != "" {
			optString = fmt.Sprintf("-%s,", o.ShortArg())
		}
		optString = fmt.Sprintf("%s--%s", optString, o.LongArg())
		m[o.LongArg()] = optString
	}
	return m
}

func (fo *FindOptions) getOptDescMap() map[string]string {
	m := map[string]string{}
	for _, o := range fo.FindOptions {
		m[o.LongArg()] = o.Description()
	}
	return m
}

func (fo *FindOptions) generateCodeFile(filePath string) {
	var buffer bytes.Buffer
	depth := 0
	buffer.WriteString("package gofind\n\n")
	buffer.WriteString("func GetFindOptions() *FindOptions {\n")
	depth++
	buffer.WriteString(fmt.Sprintf("%sreturn &FindOptions{\n", strings.Repeat("\t", depth)))
	depth++
	buffer.WriteString(fmt.Sprintf("%s[]*FindOption{\n", strings.Repeat("\t", depth)))
	depth++
	for _, o := range fo.FindOptions {
		buffer.WriteString(fmt.Sprintf("%s{\"%s\", \"%s\", \"%s\"},\n",
			strings.Repeat("\t", depth), o.ShortArg(), o.LongArg(), EscapeQuotes(o.Description())))
	}
	depth--
	buffer.WriteString(fmt.Sprintf("%s},\n", strings.Repeat("\t", depth)))
	depth--
	buffer.WriteString(fmt.Sprintf("%s}\n}\n", strings.Repeat("\t", depth)))
	err := os.WriteFile(filePath, buffer.Bytes(), 0644)
	if err != nil {
		panic(err)
	}
}
